diff --git a/tools/extra/clang-tidy/ClangTidyOptions.cpp b/tools/extra/clang-tidy/ClangTidyOptions.cpp
index c40e97c5fb6..ee5c1b31e9e 100644
--- a/tools/extra/clang-tidy/ClangTidyOptions.cpp
+++ b/tools/extra/clang-tidy/ClangTidyOptions.cpp
@@ -86,6 +86,7 @@ template <> struct MappingTraits<ClangTidyOptions> {
     bool Ignored = false;
     IO.mapOptional("Checks", Options.Checks);
     IO.mapOptional("WarningsAsErrors", Options.WarningsAsErrors);
+    IO.mapOptional("Plugins", Options.Plugins);
     IO.mapOptional("HeaderFilterRegex", Options.HeaderFilterRegex);
     IO.mapOptional("AnalyzeTemporaryDtors", Ignored); // legacy compatibility
     IO.mapOptional("FormatStyle", Options.FormatStyle);
@@ -106,6 +107,7 @@ ClangTidyOptions ClangTidyOptions::getDefaults() {
   ClangTidyOptions Options;
   Options.Checks = "";
   Options.WarningsAsErrors = "";
+  Options.Plugins = "";
   Options.HeaderFilterRegex = "";
   Options.SystemHeaders = false;
   Options.FormatStyle = "none";
@@ -145,6 +147,7 @@ ClangTidyOptions::mergeWith(const ClangTidyOptions &Other) const {
 
   mergeCommaSeparatedLists(Result.Checks, Other.Checks);
   mergeCommaSeparatedLists(Result.WarningsAsErrors, Other.WarningsAsErrors);
+  mergeCommaSeparatedLists(Result.Plugins, Other.Plugins);
   overrideValue(Result.HeaderFilterRegex, Other.HeaderFilterRegex);
   overrideValue(Result.SystemHeaders, Other.SystemHeaders);
   overrideValue(Result.FormatStyle, Other.FormatStyle);
diff --git a/tools/extra/clang-tidy/ClangTidyOptions.h b/tools/extra/clang-tidy/ClangTidyOptions.h
index 1a7535447cb..aad058ff30f 100644
--- a/tools/extra/clang-tidy/ClangTidyOptions.h
+++ b/tools/extra/clang-tidy/ClangTidyOptions.h
@@ -67,6 +67,9 @@ struct ClangTidyOptions {
   /// \brief WarningsAsErrors filter.
   llvm::Optional<std::string> WarningsAsErrors;
 
+  /// \brief List of plugins.
+  llvm::Optional<std::string> Plugins;
+
   /// \brief Output warnings from headers matching this filter. Warnings from
   /// main files will always be displayed.
   llvm::Optional<std::string> HeaderFilterRegex;
diff --git a/tools/extra/clang-tidy/tool/CMakeLists.txt b/tools/extra/clang-tidy/tool/CMakeLists.txt
index f58cfea5518..1f9ebc0c3b1 100644
--- a/tools/extra/clang-tidy/tool/CMakeLists.txt
+++ b/tools/extra/clang-tidy/tool/CMakeLists.txt
@@ -5,13 +5,14 @@ set(LLVM_LINK_COMPONENTS
   support
   )
 
-add_clang_tool(clang-tidy
+set(LLVM_OPTIONAL_SOURCES "dummy.cpp")
+add_clang_library(clangTidyMain
   ClangTidyMain.cpp
   )
-add_dependencies(clang-tidy
+add_dependencies(clangTidyMain
   clang-headers
   )
-target_link_libraries(clang-tidy
+target_link_libraries(clangTidyMain
   PRIVATE
   clangAST
   clangASTMatchers
@@ -39,11 +40,22 @@ target_link_libraries(clang-tidy
   )
 
 if(CLANG_ENABLE_STATIC_ANALYZER)
-  target_link_libraries(clang-tidy PRIVATE
+  target_link_libraries(clangTidyMain PRIVATE
     clangTidyMPIModule
   )
 endif()
 
+# supress warning about unknown source file
+set(LLVM_OPTIONAL_SOURCES "ClangTidyMain.cpp")
+# using dummy source file to supress warning about no sources
+add_clang_tool(clang-tidy
+  dummy.cpp
+  )
+target_link_libraries(clang-tidy
+  PRIVATE
+  clangTidyMain
+  )
+
 install(PROGRAMS clang-tidy-diff.py
   DESTINATION share/clang
   COMPONENT clang-tidy)
diff --git a/tools/extra/clang-tidy/tool/ClangTidyMain.cpp b/tools/extra/clang-tidy/tool/ClangTidyMain.cpp
index 12a60244f34..50e4bad96d2 100644
--- a/tools/extra/clang-tidy/tool/ClangTidyMain.cpp
+++ b/tools/extra/clang-tidy/tool/ClangTidyMain.cpp
@@ -164,6 +164,12 @@ each source file in its parent directories.
 )"),
                                    cl::init(""), cl::cat(ClangTidyCategory));
 
+static cl::opt<std::string> Plugins("plugins", cl::desc(R"(
+Comma-separated list of plugins:
+  -plugins="my-check.so,..."
+)"),
+                                   cl::init(""), cl::cat(ClangTidyCategory));
+
 static cl::opt<bool> DumpConfig("dump-config", cl::desc(R"(
 Dumps configuration in the YAML format to
 stdout. This option can be used along with a
@@ -280,6 +286,8 @@ static std::unique_ptr<ClangTidyOptionsProvider> createOptionsProvider(
     OverrideOptions.Checks = Checks;
   if (WarningsAsErrors.getNumOccurrences() > 0)
     OverrideOptions.WarningsAsErrors = WarningsAsErrors;
+  if (Plugins.getNumOccurrences() > 0)
+    OverrideOptions.Plugins = Plugins;
   if (HeaderFilter.getNumOccurrences() > 0)
     OverrideOptions.HeaderFilterRegex = HeaderFilter;
   if (SystemHeaders.getNumOccurrences() > 0)
@@ -365,6 +373,18 @@ static int clangTidyMain(int argc, const char **argv) {
   SmallString<256> FilePath = MakeAbsolute(FileName);
 
   ClangTidyOptions EffectiveOptions = OptionsProvider->getOptions(FilePath);
+  if (EffectiveOptions.Plugins) {
+    SmallVector<StringRef, 10> PluginNames;
+    StringRef(*EffectiveOptions.Plugins).split(PluginNames, ",", -1, false);
+    for (StringRef PluginNameRef : PluginNames) {
+      std::string PluginName = PluginNameRef.str();
+      std::string errMsg;
+      if (llvm::sys::DynamicLibrary::LoadLibraryPermanently(PluginName.c_str(), &errMsg)) {
+        llvm::errs() << "Failed to load plugin " << PluginName << ": " << errMsg << "\n";
+        return 1;
+      }
+    }
+  }
   std::vector<std::string> EnabledChecks =
       getCheckNames(EffectiveOptions, AllowEnablingAnalyzerAlphaCheckers);
 
diff --git a/tools/extra/clang-tidy/tool/dummy.cpp b/tools/extra/clang-tidy/tool/dummy.cpp
new file mode 100644
index 00000000000..c027d248738
--- /dev/null
+++ b/tools/extra/clang-tidy/tool/dummy.cpp
@@ -0,0 +1 @@
+typedef int dummy;
diff --git a/tools/extra/clang-tidy/tool/run-clang-tidy.py b/tools/extra/clang-tidy/tool/run-clang-tidy.py
index 93635cbef16..a33d29a4130 100755
--- a/tools/extra/clang-tidy/tool/run-clang-tidy.py
+++ b/tools/extra/clang-tidy/tool/run-clang-tidy.py
@@ -76,7 +76,7 @@ def make_absolute(f, directory):
 
 def get_tidy_invocation(f, clang_tidy_binary, checks, tmpdir, build_path,
                         header_filter, extra_arg, extra_arg_before, quiet,
-                        config):
+                        config, fix_errors):
   """Gets a command line for clang-tidy."""
   start = [clang_tidy_binary]
   if header_filter is not None:
@@ -102,6 +102,8 @@ def get_tidy_invocation(f, clang_tidy_binary, checks, tmpdir, build_path,
       start.append('-quiet')
   if config:
       start.append('-config=' + config)
+  if fix_errors:
+      start.append('-fix-errors')
   start.append(f)
   return start
 
@@ -160,7 +162,7 @@ def run_tidy(args, tmpdir, build_path, queue, lock, failed_files):
     invocation = get_tidy_invocation(name, args.clang_tidy_binary, args.checks,
                                      tmpdir, build_path, args.header_filter,
                                      args.extra_arg, args.extra_arg_before,
-                                     args.quiet, args.config)
+                                     args.quiet, args.config, args.fix_errors)
 
     proc = subprocess.Popen(invocation, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
     output, err = proc.communicate()
@@ -207,6 +209,8 @@ def main():
                       help='number of tidy instances to be run in parallel.')
   parser.add_argument('files', nargs='*', default=['.*'],
                       help='files to be processed (regex on path)')
+  parser.add_argument('-fix-errors', action='store_true', help='apply fix-its, '
+                      'even if compiler error is encountered')
   parser.add_argument('-fix', action='store_true', help='apply fix-its')
   parser.add_argument('-format', action='store_true', help='Reformat code '
                       'after applying fixes')
diff --git a/tools/extra/test/clang-tidy/check_clang_tidy.py b/tools/extra/test/clang-tidy/check_clang_tidy.py
index 9768011a3ef..7e1d1e370cf 100755
--- a/tools/extra/test/clang-tidy/check_clang_tidy.py
+++ b/tools/extra/test/clang-tidy/check_clang_tidy.py
@@ -20,6 +20,7 @@ Usage:
     [-assume-filename=<file-with-source-extension>] \
     [-check-suffix=<comma-separated-file-check-suffixes>] \
     [-check-suffixes=<comma-separated-file-check-suffixes>] \
+    [-clang-tidy=<path/to/clang-tidy>] \
     <source-file> <check-name> <temp-file> \
     -- [optional clang-tidy arguments]
 
@@ -53,6 +54,7 @@ def main():
   parser.add_argument('-check-suffix', '-check-suffixes',
                       default=[''], type=csv,
                       help="comma-separated list of FileCheck suffixes")
+  parser.add_argument('-clang-tidy', default='clang-tidy')
 
   args, extra_args = parser.parse_known_args()
 
@@ -62,6 +64,7 @@ def main():
   check_name = args.check_name
   temp_file_name = args.temp_file_name
   expect_clang_tidy_error = args.expect_clang_tidy_error
+  clang_tidy = args.clang_tidy
 
   file_name_with_extension = assume_file_name or input_file_name
   _, extension = os.path.splitext(file_name_with_extension)
@@ -138,7 +141,7 @@ def main():
   original_file_name = temp_file_name + ".orig"
   write_file(original_file_name, cleaned_test)
 
-  args = ['clang-tidy', temp_file_name, '-fix', '--checks=-*,' + check_name] + \
+  args = [clang_tidy, temp_file_name, '-fix', '--checks=-*,' + check_name] + \
         clang_tidy_extra_args
   if expect_clang_tidy_error:
     args.insert(0, 'not')
